src/ConnectedMode.UnitTests/Migration/AutoMigrationHandlerTests.cs

```diff
@@ -0,0 +1,67 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using SonarLint.VisualStudio.ConnectedMode.Migration;
+using SonarLint.VisualStudio.Core.Binding;
+using SonarLint.VisualStudio.TestInfrastructure;
+
+namespace SonarLint.VisualStudio.ConnectedMode.UnitTests.Migration
+{
+    [TestClass]
+    public class AutoMigrationHandlerTests
+    {
+        [TestMethod]
+        public void MefCtor_CheckIsExported()
+        {
+            MefTestHelpers.CheckTypeCanBeImported<AutoMigrationHandler, AutoMigrationHandler>(
+                MefTestHelpers.CreateExport<IUnintrusiveBindingController>(),
+                MefTestHelpers.CreateExport<IConfigurationProvider>(),
+                MefTestHelpers.CreateExport<IObsoleteConfigurationProvider>());
+        }
+
+        [TestMethod]
+        [DataRow(SonarLintMode.Standalone, SonarLintMode.Standalone, false)]
+        [DataRow(SonarLintMode.Connected, SonarLintMode.Standalone, true)]
+        [DataRow(SonarLintMode.LegacyConnected, SonarLintMode.Standalone, true)]
+        [DataRow(SonarLintMode.Connected, SonarLintMode.Connected, false)]
+        [DataRow(SonarLintMode.LegacyConnected, SonarLintMode.Connected, false)]
+        public void Migrate_BindingGetsCalledWithCorrectCondition(SonarLintMode obsoleteMode, SonarLintMode mode, bool expectBindingToBeCalled)
+        {
+            var bindingController = new Mock<IUnintrusiveBindingController>();
+
```

Comment:
Here I wanted to use ConfigurableConfigurationProvider but I could not make it work as I need the obsoleteConfigration Provider to be of the type IObsoleteConfigurationProvider. I still think there might be a prettier way to do this, I just do not know how.

Reply:
To be honest, just using mocks here is probably simpler and better as we only need to mock one property.

The `ConfigurationConfigurationProvider` dates from a time before we started using `Moq` (i.e. more than five years ago), so all of the `ConfigurableXXX` helpers classes are legacy code. Some of them are unnecessarily complicated e.g. the `ConfigurableConfigurationProvider` also implements `IConfigurationPersister`, which we don't need here.

My first choice when writing new tests is to use `Moq`. I only fall back on the `ConfigurableXXX` helper classes if the behaviour is hard to mock from scratch (which in turn is probably an indication that there's a design/complexity problem in the product code itself).